sherpa/tests/test_data.py

```diff
@@ -977,3 +1014,128 @@ def test_data2d_int_eval_model_to_fit(array_sizes_fixture):
+
+
+@pytest.mark.parametrize("Dataclass", REALLY_ALL_DATA_CLASSES)
+def test_data_dep_any_obj_with_mask(Dataclass):
+    args = list(INSTANCE_ARGS[Dataclass])
+    posy = POS_Y_ARRAY[Dataclass]
+    class DummyMask(list):
+        mask = 'whatisthis'
+    args[posy] = DummyMask(args[posy])
+    with pytest.warns(UserWarning, match="Set .mask"):
+        data = Dataclass(*args)
+    assert data.mask is True
+    assert len(data.get_dep(filter=True)) == len(data.get_dep(filter=False))
+
+
+# repeat set of tests except now by using ui
+# Results should be idendical, but tests are fast, so we just test again
+# To make sure that there is no heuristic in load_arrays or similar that
+# interferes with the logic
+@pytest.mark.parametrize('arrpos', [POS_X_ARRAY, POS_STATERR_ARRAY, POS_SYSERR_ARRAY])
+@pytest.mark.parametrize('Session', [Session, AstroSession])
+@pytest.mark.parametrize("data_for_load_arrays", ALL_DATA_CLASSES, indirect=True)
+def test_data_indeperr_masked_numpyarray_ui(arrpos, data_for_load_arrays, Session):
+    session, args, data = data_for_load_arrays
+    session = Session()
+    i = arrpos[type(data)]
+    mask = numpy.random.rand(*(args[i].shape)) > 0.5
+    args = list(args)
+    args[1] = numpy.ma.array(args[i], mask=mask)
+    with pytest.warns(UserWarning, match="for dependent variables only"):
+        session.load_arrays(*args)
+    new_data = session.get_data(data.name)
+    assert len(new_data.get_dep(filter=True)) == len(args[i])
+
+
+@pytest.mark.parametrize('Session', [Session, AstroSession])
+@pytest.mark.parametrize("data_for_load_arrays", ALL_DATA_CLASSES, indirect=True)
+def test_data_dep_masked_numpyarray_ui(data_for_load_arrays, Session):
+    session, args, data = data_for_load_arrays
+    session = Session()
+    posy = POS_Y_ARRAY[type(data)]
+    mask = numpy.random.rand(*(args[posy].shape)) > 0.5
+    args = list(args)
+    args[posy] = numpy.ma.array(args[posy], mask=mask)
+    session.load_arrays(*args)
+    new_data = session.get_data(data.name)
+    assert new_data.mask.shape == mask.shape
+    assert numpy.all(new_data.mask == ~mask)
+    assert len(new_data.get_dep(filter=True)) == (~mask).sum()
+
+
+@pytest.mark.parametrize('Session', [Session, AstroSession])
+@pytest.mark.parametrize("data_for_load_arrays", ALL_DATA_CLASSES, indirect=True)
+def test_data_dep_masked_numpyarray_nomask_ui(data_for_load_arrays, Session):
+    session, args, data = data_for_load_arrays
+    session = Session()
+    posy = POS_Y_ARRAY[type(data)]
+    args = list(args)
+    args[posy] = numpy.ma.array(args[posy])
+    session.load_arrays(*args)
+    new_data = session.get_data(data.name)
+    # Sherpa's way of saying "mask is not set"
+    assert new_data.mask is True
+    assert len(new_data.get_dep(filter=True)) == len(args[posy].flatten())
+
+# https://github.com/sherpa/sherpa/issues/346
+@pytest.mark.parametrize('Session', [Session, AstroSession])
+def test_regression_346(Session):
```

Comment:
Not issue #346, but what happens if you try to create a Data object (either directly or via load_arrays) where all points are masked out? It looks like `Data1D` at least supports this, so I imagine it'll result in something like the following, but wanted to check:

```
In [20]: from sherpa.data import Data1D

In [21]: d = Data1D('x', [], [])
/home/dburke/sherpa/sherpa-master/sherpa/data.py:49: UserWarning: Converting array [] to numpy array
  warnings.warn("Converting array {} to numpy array".format(array))

In [22]: print(d)
name      = x
x         = Float64[0]
y         = Float64[0]
staterror = None
syserror  = None
```

Reply:
@hamogu - do you know what happens in this case ? I didn't see any obvious test changes that exercised this, but I could well have missed it.

Reply:
It behaves as it should. Your code above works exactly as before because it passes in lists, not numpy arrays and that behaviour is unchanged. If you use masked array, you can mask as much as you like. The masked datapoints are there and can be retried just as you can do after using `set_mask`. That behaviour is unchanged, too. All the changes is that instead of (assuming y is a numpy masked array)
```
d = Data1D('x', x, y)
d.set_mask(~y.mask)
```
you can now simply do `d=Data1D('x', x, y)`. It really just saves you that one line of typing. This PR is important not because it helps Sherpa do something it could not do before, but because is aligns the default to what users apparently expect (#346, #695) - and that expectation "Don't just ignore masks without even telling me you do that" is totally reasonable.

Full output from the tests @DougBurke asked for:
```
sherpa In [1]: from sherpa.data import Data1D 
          ...:                                                                                                       

sherpa In [2]: d = Data1D('x', [], [])                                                                               
/Users/guenther/anaconda3/envs/sherpaciao/lib/python3.7/site-packages/sherpa-4.12.1+301.gecfad5f4.dirty-py3.7-macosx-10.9-x86_64.egg/sherpa/data.py:49: UserWarning: Converting array [] to numpy array.
  warnings.warn("Converting array {} to numpy array.".format(array))

sherpa In [3]: print(d)                                                                                              
name      = x
x         = Float64[0]
y         = Float64[0]
staterror = None
syserror  = None

sherpa In [6]: xarr = numpy.ma.array([1,2,3], mask=[True, True, True])                                               

sherpa In [7]: yarr = [1, 2, 1]                                                                                      

sherpa In [10]: d = Data1D('x', xarr, yarr)                                                                          

sherpa In [11]: print(d)                                                                                             
name      = x
x         = Int64[3]
y         = Int64[3]
staterror = None
syserror  = None

sherpa In [12]: d.get_indep()                                                                                        
Out[12]: (array([1, 2, 1]),)

sherpa In [13]: d.get_indep(filter=True)                                                                             
Out[13]: (array([], dtype=int64),)
```

Reply:
Thanks. It's an edge case to check, but someone always finds them ;-)

Reply:
I'll leave it up to you to see if it needs an explicit test or not.